README.md
# vscode-color-exchange README
![Build Status](https://travis-ci.org/IronLu233/vscode-color-exchange.svg?branch=master)
[![codecov](https://codecov.io/gh/IronLu233/vscode-color-exchange/branch/master/graph/badge.svg)](https://codecov.io/gh/IronLu233/vscode-color-exchange)

A VS code extension uses `VS code`'s code action for refactoring your color unit code. It can transform your color unit to RGB, RGBA, etc.

## Features
convert your color token to:
- RGB, RGBA
- HSL
- 6-dight hexadecimal and 3-dight hexadecimal(if available)

![2018-10-03 11 59 49](https://user-images.githubusercontent.com/20639676/46391328-f04f5280-c70e-11e8-8fb9-fe68ae643f8c.gif)
## Installation
In order to install an extension you need to launch the Command Pallete (Ctrl + Shift + P or Cmd + Shift + P) and type Extensions. There you have either the option to show the already installed snippets or install new ones. Search for `color-exchange` and install it.

## Usage
After installation and reload VS code, open any file and move the cursor into color unit, click the lightbulbs in your editor or Use shortcut: `Cmd + .`(`Ctrl + .` in windows), then select your new color unit.

## Requirements
`vs code` >=1.27.0

## TODOS
- [ ] Add more unit test
- [x] Support convert other color to keyword color if available
- [x] Add CI
- [x] Add `codecov` for code coverage
<!-- ## Known issues
- Code action will appear at incorrect place, when cursor after `;` in `rgb(255, 255, 255);` -->

## Release Notes
### 1.1.0
- Add support for convert to keyword color if available

### 1.0.1
- Fix code action incorrect appeared when cursor in the tail of line.
- Fix code action not appeared when cursor in color unit's function notation
- Fix the same color token in code actions which is need refactor.
### 1.0.0
Basic feature.

-----------------------------------------------------------------------------------------------------------
**Enjoy!**


CHANGELOG.md
# Change Log
All notable changes to the "vscode-color-exchange" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]
- Initial release

test/index.ts
//
// PLEASE DO NOT MODIFY / DELETE UNLESS YOU KNOW WHAT YOU ARE DOING
//
// This file is providing the test runner to use when running extension tests.
// By default the test runner in use is Mocha based.
//
// You can provide your own test runner if you want to override it by exporting
// a function run(testRoot: string, clb: (error:Error) => void) that the extension
// host can call to run the tests. The test runner is expected to use console.log
// to report the results back to the caller. When the tests are finished, return
// a possible error to the callback or null if none.

import * as testRunner from './testRunner';

// You can directly control Mocha options by uncommenting the following lines
// See https://github.com/mochajs/mocha/wiki/Using-mocha-programmatically#set-options for more info
testRunner.configure({
    ui: 'tdd', 		// the TDD UI is being used in extension.test.ts (suite, test, etc.)
    useColors: true // colored output from test results
});

module.exports = testRunner;

test/color.test.ts
import * as assert from 'assert';


import Color from "../src/color";
import keyword2HexadecimalMap from '../src/keyword2HexadecimalMap';
import hexadecimal2Keyword from '../src/hexadecimal2KeywordMap';

// TODO: add situations in https://test.csswg.org/harness/test/css-color-3_dev/section to unit test case
suite('test color model', () => {
  suite('convert from keyword token', () => {
    test('color transparent', () => {
      const transparentColor = Color.fromKeywordColorToken('transparent') as Color;
      assert.equal(transparentColor.transparent, 'transparent');
    });

    test('3-dight hexadecimal color', () => {
      const color = Color.fromKeywordColorToken('#ABC') as Color;
      assert.equal(color.sixDightHexadecimal, '#aabbcc');
      assert.equal(color.threeDightHexadecimal, '#abc');
      assert.equal(color.RGB, 'rgb(170, 187, 204)');
      assert.equal(color.RGBA, 'rgba(170, 187, 204, 1)');
      assert.equal(color.HSL, 'hsl(210, 25%, 73%)');
    });

    test('6-dight hexadecimal color #1', () => {
      const color = Color.fromKeywordColorToken('#ABCDEF') as Color;
      assert.equal(color.threeDightHexadecimal, '');
      assert.equal(color.sixDightHexadecimal, '#abcdef');
      assert.equal(color.RGB, 'rgb(171, 205, 239)');
      assert.equal(color.RGBA, 'rgba(171, 205, 239, 1)');
      assert.equal(color.HSL, 'hsl(210, 68%, 80%)');
    });

    test('6-dight hexadecimal color #2', () => {
      const color = Color.fromKeywordColorToken('#AABBCC') as Color;
      assert.equal(color.threeDightHexadecimal, '#abc');
      assert.equal(color.sixDightHexadecimal, '#aabbcc');
      assert.equal(color.RGB, 'rgb(170, 187, 204)');
      assert.equal(color.RGBA, 'rgba(170, 187, 204, 1)');
      assert.equal(color.HSL, 'hsl(210, 25%, 73%)');
    });

    test('convert to keyword color unit', () => {
      Object.keys(hexadecimal2Keyword).forEach(hexadecimal => {
        const color = Color.fromKeywordColorToken(hexadecimal);
        assert.notEqual(color, null);
        assert.equal((color as Color).keyword, hexadecimal2Keyword[hexadecimal]);
      });
    });

    test('valid keyword color', () => {
      Object.keys(keyword2HexadecimalMap).forEach(key => {
        const color = Color.fromKeywordColorToken(key) as Color;
        assert.equal(color.sixDightHexadecimal.toUpperCase(), keyword2HexadecimalMap[key]);
      });
    });

    test('invalid keyword color', () => {
      assert.equal(Color.fromKeywordColorToken('invalidToken'), null);
    });

    test('invalid hexadecimal color #1', () => {
      assert.equal(Color.fromKeywordColorToken('#AACF'), null);
    });

    test('invalid hexadecimal color #2', () => {
      assert.equal(Color.fromKeywordColorToken('#GGZZWW'), null);
    });
  });

  suite('convert from functional color', () => {
    test('RGB color #1', () => {
      const color = Color.fromFunctionalColorToken('rgb(255, 255, 255)') as Color;
      assert.notEqual(color, null);
      assert.equal(color.RGB, 'rgb(255, 255, 255)');
      assert.equal(color.RGBA, 'rgba(255, 255, 255, 1)');
      assert.equal(color.percentRGB, 'rgb(100.0%, 100.0%, 100.0%)');
      assert.equal(color.percentRGBA, 'rgba(100.0%, 100.0%, 100.0%, 100.0%)');
      assert.equal(color.sixDightHexadecimal, '#ffffff');

    });
    test('RGB color #2', () => {
      const color = Color.fromFunctionalColorToken('rgb(50%, 50%, 50%)') as Color;

      assert.notEqual(color, null);
      assert.equal(color.RGB, 'rgb(128, 128, 128)');
      assert.equal(color.percentRGB, 'rgb(50.2%, 50.2%, 50.2%)');
      assert.equal(color.RGBA, 'rgba(128, 128, 128, 1)');
      assert.equal(color.percentRGBA, 'rgba(50.2%, 50.2%, 50.2%, 100.0%)');
    });

    test('RGBA color #1', () => {
      const color = Color.fromFunctionalColorToken('rgba(128, 128, 128, .5)') as Color;
      assert.notEqual(color, null);
      assert.equal(color.RGB, '');
      assert.equal(color.RGBA, 'rgba(128, 128, 128, 0.5)');
      assert.equal(color.percentRGB, '');
      assert.equal(color.percentRGBA, 'rgba(50.2%, 50.2%, 50.2%, 50.0%)');
    });

    test('RGBA color #2', () => {
      const color = Color.fromFunctionalColorToken('rgba(50%, 50%, 50%, 50%)') as Color;

      assert.notEqual(color, null);
      assert.equal(color.RGB, '');
      assert.equal(color.percentRGB, '');
      assert.equal(color.RGBA, 'rgba(128, 128, 128, 0.5)');
      assert.equal(color.percentRGBA, 'rgba(50.2%, 50.2%, 50.2%, 50.0%)');
    });

    test('HSL color #1', () => {
      assert.deepEqual(
        Color.fromFunctionalColorToken('hsl(240, 100%, -100%)'),
        Color.fromKeywordColorToken('black')
      );
    });

    test('HSL color #2', () => {
      assert.deepEqual(
        Color.fromFunctionalColorToken('hsl(240, 75%, -20%)'),
        Color.fromKeywordColorToken('black')
      );
    });

    test('HSL color #3', () => {
      assert.deepEqual(
        Color.fromFunctionalColorToken('hsl(240, 75%, 120%)'),
        Color.fromKeywordColorToken('white')
      );
    });

    test('HSL color #4', () => {

      assert.deepEqual(
        Color.fromFunctionalColorToken('hsl(240, 130%, 50%)'),
        Color.fromFunctionalColorToken('rgb(0, 0, 255)')
      );
    });

    test('HSL color #5', () => {
      assert.deepEqual(
        Color.fromFunctionalColorToken('hsl(264, 100%, 50%)'),
        Color.fromFunctionalColorToken('rgb(102, 0, 255)')
      );
    });

    test('HSL color #6', () => {
      assert.deepEqual(
        Color.fromFunctionalColorToken('hsl(0, -50%, 40%)'),
        Color.fromFunctionalColorToken('rgb(102, 102, 102)')
      );
    });

    test('HSL color #7', () => {
      assert.deepEqual(
        Color.fromFunctionalColorToken('hsl(30, -50%, 60%)'),
        Color.fromFunctionalColorToken('rgb(153, 153, 153)')
      );
    });
  });
});


test/testRunner.ts
/*---------------------------------------------------------
 * Copyright (C) Microsoft Corporation. All rights reserved.
 *--------------------------------------------------------*/

'use strict';

import * as fs from "fs";
import * as paths from 'path';
import * as glob from 'glob';
const istanbul = require("istanbul");
const remapIstanbul = require("remap-istanbul");

// Linux: prevent a weird NPE when mocha on Linux requires the window size from the TTY
// Since we are not running in a tty environment, we just implementt he method statically
var tty = require('tty');
if (!tty.getWindowSize) {
  tty.getWindowSize = function () { return [80, 75]; };
}

import Mocha = require('mocha');

let mocha = new Mocha(<any>{
  ui: 'tdd',
  useColors: true
});


export function configure(opts: any): void {
  mocha = new Mocha(opts);
}

function _mkDirIfExists(dir: string): void {
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir);
  }
}

function _readCoverOptions(testsRoot: string): ITestRunnerOptions | undefined {
  const coverConfigPath = paths.join(testsRoot, "..", "..", "coverconfig.json");
  let coverConfig: ITestRunnerOptions | undefined = undefined;
  if (fs.existsSync(coverConfigPath)) {
    let configContent = fs.readFileSync(coverConfigPath, "utf-8");
    coverConfig = JSON.parse(configContent);
  }
  return coverConfig;
}


export function run(testsRoot: string, clb: (error, failures?: number) => void): void {

  // Enable source map support
  require('source-map-support').install();

  const coverOptions = _readCoverOptions(testsRoot);
  if (coverOptions && coverOptions.enabled) {
    // Setup coverage pre-test, including post-test hook to report
    let coverageRunner = new CoverageRunner(coverOptions, testsRoot, clb);
    coverageRunner.setupCoverage();
  }

  // Glob test files
  glob('**/**.test.js', { cwd: testsRoot }, (error, files) => {
    if (error) {
      return clb(error);
    }

    try {

      // Fill into Mocha
      files.forEach(f => mocha.addFile(paths.join(testsRoot, f)));

      // Run the tests
      mocha.run((failures) => {
        clb(null, failures);
      });
    } catch (error) {
      return clb(error);
    }
  });
}

interface ITestRunnerOptions {
  enabled?: boolean;
  relativeCoverageDir: string;
  relativeSourcePath: string;
  ignorePatterns: string[];
  includePid?: boolean;
  reports?: string[];
  verbose?: boolean;
}


class CoverageRunner {

  private coverageVar: string = "$$cov_" + new Date().getTime() + "$$";
  private transformer: any = undefined;
  private matchFn: any = undefined;
  private instrumenter: any = undefined;

  constructor(private options: ITestRunnerOptions, private testsRoot: string, private endRunCallback: any) {
    if (!options.relativeSourcePath) {
      return endRunCallback("Error - relativeSourcePath must be defined for code coverage to work");
    }
  }

  public setupCoverage(): void {
    // Set up Code Coverage, hooking require so that instrumented code is returned
    let self = this;
    self.instrumenter = new istanbul.Instrumenter({ coverageVariable: self.coverageVar });
    let sourceRoot = paths.join(self.testsRoot, self.options.relativeSourcePath);

    // Glob source files
    let srcFiles = glob.sync('**/**.js', {
      cwd: sourceRoot,
      ignore: self.options.ignorePatterns,
    });

    // Create a match function - taken from the run-with-cover.js in istanbul.
    let decache = require("decache");
    let fileMap = {};
    srcFiles.forEach((file) => {
      let fullPath = paths.join(sourceRoot, file);
      fileMap[fullPath] = true;

      // On Windows, extension is loaded pre-test hooks and this mean we lose
      // our chance to hook the Require call. In order to instrument the code
      // we have to decache the JS file so on next load it gets instrumented.
      // This doesn"t impact tests, but is a concern if we had some integration
      // tests that relied on VSCode accessing our module since there could be
      // some shared global state that we lose.
      decache(fullPath);
    });

    self.matchFn = (file): boolean => { return fileMap[file]; };
    self.matchFn.files = Object.keys(fileMap);

    // Hook up to the Require function so that when this is called, if any of our source files
    // are required, the instrumented version is pulled in instead. These instrumented versions
    // write to a global coverage variable with hit counts whenever they are accessed
    self.transformer = self.instrumenter.instrumentSync.bind(self.instrumenter);
    let hookOpts = { verbose: false, extensions: [".js"] };
    istanbul.hook.hookRequire(self.matchFn, self.transformer, hookOpts);

    // initialize the global variable to stop mocha from complaining about leaks
    global[self.coverageVar] = {};

    // Hook the process exit event to handle reporting
    // Only report coverage if the process is exiting successfully
    process.on("exit", (code) => {
      self.reportCoverage();
    });
  }

  /**
   * Writes a coverage report. Note that as this is called in the process exit callback, all calls must be synchronous.
   *
   * @returns {void}
   *
   * @memberOf CoverageRunner
   */
  public reportCoverage(): void {
    let self = this;
    istanbul.hook.unhookRequire();
    let cov: any;
    if (typeof global[self.coverageVar] === "undefined" || Object.keys(global[self.coverageVar]).length === 0) {
      console.error("No coverage information was collected, exit without writing coverage information");
      return;
    } else {
      cov = global[self.coverageVar];
    }

    // TODO consider putting this under a conditional flag
    // Files that are not touched by code ran by the test runner is manually instrumented, to
    // illustrate the missing coverage.
    self.matchFn.files.forEach((file) => {
      if (!cov[file]) {
        self.transformer(fs.readFileSync(file, "utf-8"), file);

        // When instrumenting the code, istanbul will give each FunctionDeclaration a value of 1 in coverState.s,
        // presumably to compensate for function hoisting. We need to reset this, as the function was not hoisted,
        // as it was never loaded.
        Object.keys(self.instrumenter.coverState.s).forEach((key) => {
          self.instrumenter.coverState.s[key] = 0;
        });

        cov[file] = self.instrumenter.coverState;
      }
    });

    // TODO Allow config of reporting directory with
    let reportingDir = paths.join(self.testsRoot, self.options.relativeCoverageDir);
    let includePid = self.options.includePid;
    let pidExt = includePid ? ("-" + process.pid) : "";
    let coverageFile = paths.resolve(reportingDir, "coverage" + pidExt + ".json");

    _mkDirIfExists(reportingDir); // yes, do this again since some test runners could clean the dir initially created

    fs.writeFileSync(coverageFile, JSON.stringify(cov), "utf8");

    let remappedCollector = remapIstanbul.remap(cov, {
      warn: warning => {
        // We expect some warnings as any JS file without a typescript mapping will cause this.
        // By default, we"ll skip printing these to the console as it clutters it up
        if (self.options.verbose) {
          console.warn(warning);
        }
      }
    });

    let reporter = new istanbul.Reporter(undefined, reportingDir);
    let reportTypes = (self.options.reports instanceof Array) ? self.options.reports : ["lcov"];
    reporter.addAll(reportTypes);
    reporter.write(remappedCollector, true, () => {
      console.log(`reports written to ${reportingDir}`);
    });
  }
}

src/extension.ts
'use strict';
import * as vscode from 'vscode';
import { window, Command, Position, WorkspaceEdit, workspace, Uri, Range, TextLine } from 'vscode';
import Color from './color';
import { getKeywordColorToken, getFunctionalColorToken } from './getColorToken';

// NOTE: use `import` will leave an error of leaking type definition
const LRUMap = require('collections/lru-map');

const cache = LRUMap({}, 500);

export function activate(context: vscode.ExtensionContext) {

    console.log('"vscode-color-exchange" is active!');

    let disposable = vscode.commands.registerCommand('color-exchange.changeColor',
        (uri: Uri, colorUnit: string, newText: string, cursorPosition: Position, line: TextLine) => {
            const tokenStartCharacterIndex = line.text.indexOf(colorUnit);

            if (!window.activeTextEditor || tokenStartCharacterIndex === -1) { return; }

            const startPosition = new Position(cursorPosition.line, tokenStartCharacterIndex);
            const endPosition = startPosition.translate(0, colorUnit.length);

            const editRange = new Range(startPosition, endPosition);
            const edit = new WorkspaceEdit();
            edit.replace(uri, editRange, newText);
            workspace.applyEdit(edit);
        }
    );

    context.subscriptions.push(disposable);

    disposable = vscode.languages.registerCodeActionsProvider('*', {
        provideCodeActions(document) {
            const commands: Command[] = [];
            const { activeTextEditor } = window;

            if (!activeTextEditor) {
                return [];
            }

            const { start: startPosition } = activeTextEditor.selection;
            const line = document.lineAt(startPosition);

            function pushModifyCommand(token: string, newText: string) {
                if (token === newText) { return; }
                if (newText) {
                    commands.push({
                        title: newText,
                        command: 'color-exchange.changeColor',
                        arguments: [document.uri, token, newText, startPosition, line]
                    });
                }
            }

            let token: string, color: Color | null = null;

            if (token = getKeywordColorToken(line.text, startPosition.character)) {
                color = cache.get(token) || Color.fromKeywordColorToken(token);
            }
            if (!color && (token = getFunctionalColorToken(line.text, startPosition.character))) {
                color = cache.get(token) || Color.fromFunctionalColorToken(token);
            }

            if (!color) {
                // this line has no token, return empty array
                return [];
            }

            // set the cache
            if (!cache.has(token)) {
                cache.set(token, color);
            }

            const {
                threeDightHexadecimal,
                sixDightHexadecimal,
                RGB,
                percentRGB,
                RGBA,
                percentRGBA,
                HSL,
                keyword
            } = color;
            [
                threeDightHexadecimal,
                threeDightHexadecimal.toUpperCase() !== threeDightHexadecimal ?
                    threeDightHexadecimal.toUpperCase() : '',
                sixDightHexadecimal,
                sixDightHexadecimal.toUpperCase() !== sixDightHexadecimal ?
                    sixDightHexadecimal.toUpperCase() : '',
                RGB,
                percentRGB,
                RGBA,
                percentRGBA,
                HSL,
                keyword
            ].forEach(colorUnit => pushModifyCommand(token, colorUnit));

            return commands;
        }
    });

    context.subscriptions.push(disposable);
}

export function deactivate() {
}

src/getColorToken.ts
const keywordColorTokenPrefixRegex = /#?[\w]*$/;
const keywordColorTokenSuffixRegex = /^#?[\w]*/;

const functionalColorTokenPrefixRegex = /\(.*?$/;
const functionalColorTokenSuffixRegex = /^.*?\)/;
const roughFunctionalColorRegex = /((rgba)|(rgb)|(hsl))\(.+?\)/i;
const functionNotationRegex = /(rgba)|(rgb)|(hsl)/i;

export function getKeywordColorToken(line: string, cursor: number): string {
  const prefix = line.substr(0, cursor);
  const suffix = line.substr(cursor);
  const prefixMatchArray = keywordColorTokenPrefixRegex.exec(prefix);
  const suffixMatchArray = keywordColorTokenSuffixRegex.exec(suffix);
  let resultToken = '';
  if (prefixMatchArray) {
    resultToken += prefixMatchArray[0];
  }
  if (suffixMatchArray) {
    resultToken += suffixMatchArray[0];
  }
  return resultToken;
}

export function getFunctionalColorToken(line: string, cursor: number): string {

  const prefix = line.substr(0, cursor);
  const suffix = line.substr(cursor);
  const prefixMatchArray = functionalColorTokenPrefixRegex.exec(prefix);
  const suffixMatchArray = functionalColorTokenSuffixRegex.exec(suffix);
  let resultToken = '';
  if (prefixMatchArray) { // matched left parenthesis
    if (!suffixMatchArray) { // all token string is on the left of cursor
      const tokenContainsString = line.slice(
        Math.max(prefixMatchArray.index - 4, 0),
        prefixMatchArray.index + prefixMatchArray[0].length
      );
      const roughFunctionalColorMatchArray = roughFunctionalColorRegex.exec(tokenContainsString);

      if (!roughFunctionalColorMatchArray) { return ''; }
      const tokenStartIndex = Math.max(prefixMatchArray.index - 4, 0) + roughFunctionalColorMatchArray.index;
      const tokenEndIndex = tokenStartIndex + roughFunctionalColorMatchArray[0].length;
      return tokenEndIndex >= cursor ? roughFunctionalColorMatchArray[0] : '';
    }

    resultToken += prefixMatchArray[0];

    const functionNotation = prefix.substr(Math.max(prefixMatchArray.index - 4, 0), prefixMatchArray.index);
    const functionNotationMatchArray = functionNotationRegex.exec(functionNotation);
    if (functionNotationMatchArray) {
      resultToken = functionNotationMatchArray[0] + resultToken;
    } else {
      // the word before left parenthesis is not 'rgb' or 'rgba' or 'hsl'
      // means it doesn't match any functional color
      return '';
    }

  }
  if (suffixMatchArray) { // matched right parenthesis
    if (!prefixMatchArray) {
      const tokenContainsString = line.substr(
        Math.max(0, prefix.length + suffixMatchArray.index - 4),
        prefix.length + suffixMatchArray.index + suffixMatchArray[0].length
      );

      const roughFunctionalColorMatchArray = roughFunctionalColorRegex.exec(tokenContainsString);
      if (!roughFunctionalColorMatchArray) { return ''; }
      const tokenStartIndex = Math.max(0, prefix.length + suffixMatchArray.index - 4) + roughFunctionalColorMatchArray.index;
      return tokenStartIndex <= cursor ? roughFunctionalColorMatchArray[0] : '';
    }

    resultToken += suffixMatchArray[0];
  }

  return resultToken;
}

src/color.ts
import keyword2HexadecimalMap from "./keyword2HexadecimalMap";
import hexadecimal2Keyword from "./hexadecimal2KeywordMap";

const hexadecimalRegex = /^#([0-9a-f]{3}|[0-9a-f]{6})$/i;
const canTransformToThreeDightHexadacimalRegex = /#(?:([0-9a-f])\1){3}/i;

const functionalColorRegex = /((?:RGBA)|(?:RGB)|(?:HSL))\((?:\s*(-?[\d\.%]+)\s*,\s*(-?[\d\.%]+)\s*,\s*(-?[\d\.%]+)\s*(?:,\s*(-?[\d\.%]+)\s*)?)\)/i;

const RGBAVectorTokenRegex = /((?:\d*\.\d+)|(?:\d+))(%)?/;

const HueVectorTokenRegex = /\d+/;
const SaturationLightnessVectorTokenRegex = /(-?\d+)%/;


export function RGB2HSL(red: number, green: number, blue: number) {
  red /= 255;
  green /= 255;
  blue /= 255;
  const [max, min] = [red, green, blue].reduce(
    ([prevMax, prevMin], current: number): [number, number] =>
      [
        Math.max(prevMax, current),
        Math.min(prevMin, current)
      ],
    [0, 1]);
  const lightness = (max + min) / 2;
  const delta = (max - min);
  let hue = 0, saturation = 0;

  switch (max) {
    case min:
      hue = 0;
      break;
    case red:
      hue = 60 * ((green - blue) / delta) + (green >= blue ? 0 : 360);
      break;
    case green:
      hue = 60 * ((blue - red) / delta) + 120;
      break;
    case blue:
      hue = 60 * ((red - green) / delta) + 240;
      break;
  }

  if (lightness === 0 || max === min) {
    saturation = 0;
  } else if (0 < lightness && lightness <= 1 / 2) {
    saturation = delta / (2 * lightness);
  } else if (lightness > 1 / 2) {
    saturation = delta / (2 - 2 * lightness);
  }

  return [hue, saturation, lightness];
}

export function HSL2RGB(hue: number, saturation: number, lightness: number) {
  // normalize hue angle
  hue = ((hue % 360) + 360) % 360;
  if (saturation === 0) {
    return [lightness, lightness, lightness].map(v => Math.floor(v * 255));
  }

  const q = lightness < 0.5 ?
    lightness * (1 + saturation) :
    lightness + saturation - (lightness * saturation);
  const p = 2 * lightness - q;
  const hk = hue / 360;
  return [
    hk + 1 / 3,
    hk,
    hk - 1 / 3
  ]
    .map(tc => {
      if (tc < 0) {
        return tc + 1;
      }
      if (tc > 1) {
        return tc - 1;
      }
      return tc;
    })
    .map(
      tc => {
        if (tc < 1 / 6) {
          return p + ((q - p) * 6 * tc);
        }
        if (1 / 6 <= tc && tc < 1 / 2) {
          return q;
        }
        if (1 / 2 <= tc && tc < 2 / 3) {
          return p + ((q - p) * 6 * (2 / 3 - tc));
        }
        return p;
      }
    )
    .map(tc => Math.round(parseFloat(tc.toFixed(8)) * 255));
}


export default class Color {
  public static fromKeywordColorToken(token: string): Color | null {
    if (token === 'transparent') {
      return new Color(0, 0);
    }

    // hexadecimal token
    if (token.startsWith('#')) {
      if (!hexadecimalRegex.test(token)) {
        return null;
      }

      return new Color(token, 1);

    }

    const hexadecimal = keyword2HexadecimalMap[token];
    if (hexadecimal) {
      return new Color(hexadecimal, 1);
    }

    return null;
  }

  public static fromFunctionalColorToken(token: string): Color | null {
    const matchedArray = functionalColorRegex.exec(token);
    if (!matchedArray) { return null; }

    const [, schema, ...vectorTokens] = matchedArray;
    if (schema.toLowerCase() === 'rgb' || schema.toLowerCase() === 'rgba') {
      let [redToken, greenToken, blueToken, alphaToken = '1'] = vectorTokens;
      const tokenRegexMatchArrays = [redToken, greenToken, blueToken]
        .map(token => RGBAVectorTokenRegex.exec(token));
      if (tokenRegexMatchArrays.some(regex => !regex)) {
        return null;
      }
      const RGBVectors: number[] = [];
      for (let matchArray of tokenRegexMatchArrays) {
        // just for passing typescript language type check
        if (!matchArray) {
          return null;
        }

        let [, valueToken, percentage] = matchArray;
        if (valueToken.startsWith('.')) {
          valueToken = '0' + valueToken;
        }

        let value = Math.round(percentage ?
          parseFloat(valueToken) * 255 / 100 :
          parseInt(valueToken));

        if (isNaN(value) || value >> 24) {
          return null;
        }

        value = Math.max(value, 0);
        value = Math.min(value, 255);

        RGBVectors.push(value);
      }

      const alphaMatchArray = RGBAVectorTokenRegex.exec(alphaToken);
      let alphaValue = 1;
      if (alphaMatchArray) {
        let [, valueToken, percentage] = alphaMatchArray;
        if (valueToken.startsWith('.')) {
          valueToken = '0' + valueToken;
        }

        alphaValue = percentage ?
          parseFloat(valueToken) / 100 :
          parseFloat(valueToken);

        alphaValue = Math.max(0, alphaValue);
        alphaValue = Math.min(1, alphaValue);
      }

      return new Color(RGBVectors, alphaValue, 'rgb');
    } else if (schema === 'hsl') {
      const [hueToken, saturationToken, lightnessToken] = vectorTokens;
      const HSLVectors: number[] = [];

      // use `test` instead of `exec`, matched string is our expected

      if (!HueVectorTokenRegex.test(hueToken)) { return null; }
      let hueValue = parseInt(hueToken);

      HSLVectors.push(hueValue);

      const matchArrays = [saturationToken, lightnessToken]
        .map(token => SaturationLightnessVectorTokenRegex.exec(token));

      if (matchArrays.some(array => !array)) {
        return null;
      }
      for (let matchArray of matchArrays) {
        // just for passing Typescript type check
        if (!matchArray) {
          return null;
        }
        const [, valueToken] = matchArray;
        let value = parseInt(valueToken) / 100;
        value = Math.min(1, value);
        value = Math.max(0, value);
        HSLVectors.push(value);
      }
      return new Color(HSLVectors, 1, 'hsl');
    }
    return null;
  }

  private constructor(
    colorValue: string | number | number[],
    alpha = 1,
    schema: 'rgb' | 'hsl' | 'hexadecimal' = 'hexadecimal'
  ) {
    this.alpha = alpha;

    if (schema === 'hexadecimal' && !Array.isArray(colorValue)) {
      this.initHexadecimalColorValue(colorValue);
    }

    if (Array.isArray(colorValue)) {
      if (schema === 'rgb') {
        this.initRGBColorValue(colorValue);
      } else if (schema === 'hsl') {
        this.initHSLColorValue(colorValue);
      }
    }

  }

  private initHexadecimalColorValue(colorValue: string | number) {
    if (typeof colorValue === 'string') {
      colorValue = colorValue.toLowerCase();
      this._threeDigitHexadecimal = colorValue.length === 4 ? colorValue : '';

      this._sixDightHexadecimal = colorValue.length === 4 ?
        '#' + colorValue
          .slice(1)
          .split("")
          .map(dight => dight.repeat(2))
          .join("")
        : colorValue;

      this.hexadecimalValue = parseInt(this.sixDightHexadecimal.replace('#', '0x'));
    } else {
      this.hexadecimalValue = colorValue;
      this.setSixDightHexadecimal();
    }

    this.setRGBAndRGBAColor();
    this.setHSLColor();
    this.setThreeDightHexadecimal();
  }

  private initRGBColorValue([red, green, blue]: number[]) {
    // #FFFFFF
    this.hexadecimalValue = (red << 16) | (green << 8) | blue;
    this.setSixDightHexadecimal();
    this.setThreeDightHexadecimal();
    this.setRGBAndRGBAColor();
    this.setHSLColor();
  }

  private initHSLColorValue([hue, saturation, lightness]: number[]) {
    this.setHSLColor(hue, saturation, lightness);
    this.initRGBColorValue(HSL2RGB(hue, saturation, lightness));
  }

  private setSixDightHexadecimal() {
    const rowHexadecimal = this.hexadecimalValue.toString(16);
    this._sixDightHexadecimal = `#${'0'.repeat(6 - rowHexadecimal.length)}${rowHexadecimal}`;
  }

  private setRGBAndRGBAColor() {
    if (this.alpha === 0) {
      this._RGBA = 'rgba(0, 0, 0, 0)';
      this._percentRGBA = 'rgba(0%, 0%, 0%, 0)';
      return;
    }

    if (this.RGB === '' && this.RGBA === '') {
      const blue = this.hexadecimalValue & 0x0000FF;
      const green = (this.hexadecimalValue & 0x00FF00) >> 8;
      const red = (this.hexadecimalValue & 0xFF0000) >> 16;

      const redPercent = (red / 255 * 100).toFixed(1);
      const greenPercent = (green / 255 * 100).toFixed(1);
      const bluePercent = (blue / 255 * 100).toFixed(1);
      const alphaPercent = (this.alpha * 100).toFixed(1);

      if (this.alpha === 1) {
        this._RGB = `rgb(${red}, ${green}, ${blue})`;
        this._percentRGB = `rgb(${redPercent}%, ${greenPercent}%, ${bluePercent}%)`;
      }
      this._RGBA = `rgba(${red}, ${green}, ${blue}, ${this.alpha})`;
      this._percentRGBA = `rgba(${redPercent}%, ${greenPercent}%, ${bluePercent}%, ${alphaPercent}%)`;
    }
  }

  private setThreeDightHexadecimal() {
    if (this.threeDightHexadecimal === '' && this.alpha === 1) {
      if (canTransformToThreeDightHexadacimalRegex.test(this.sixDightHexadecimal)) {

        const blueLowDight = (this.hexadecimalValue & 0x00000F).toString(16);
        const greenLowDight = ((this.hexadecimalValue & 0x000F00) >> 8).toString(16);
        const redLowDight = ((this.hexadecimalValue & 0x0F0000) >> 16).toString(16);

        this._threeDigitHexadecimal = `#${redLowDight}${greenLowDight}${blueLowDight}`;
      } else {
        this._threeDigitHexadecimal = '';
      }
    }
  }

  private setHSLColor(...params: number[]) {
    if (params.length === 0) {
      const blue = this.hexadecimalValue & 0x0000FF;
      const green = (this.hexadecimalValue & 0x00FF00) >> 8;
      const red = (this.hexadecimalValue & 0xFF0000) >> 16;
      params = RGB2HSL(red, green, blue);
    }

    let [hue, saturation, lightness] = params;

    // normalize hue angle
    hue = ((hue % 360) + 360) % 360;
    this._HSL = `hsl(${Math.round(hue)}, ${(saturation * 100).toFixed()}%, ${(lightness * 100).toFixed()}%)`;

  }

  hexadecimalValue = 0;
  private alpha = 1;

  private _sixDightHexadecimal = '';

  get sixDightHexadecimal() {
    return this._sixDightHexadecimal;
  }

  private _threeDigitHexadecimal = '';
  get threeDightHexadecimal() {
    return this._threeDigitHexadecimal;
  }

  private _RGB = '';

  get RGB() {
    return this._RGB;
  }

  private _percentRGB = '';
  get percentRGB() {
    return this._percentRGB;
  }

  private _RGBA = '';

  get RGBA() {
    return this._RGBA;
  }

  private _percentRGBA = '';
  get percentRGBA() {
    return this._percentRGBA;
  }

  private _HSL = '';

  get HSL() {
    return this._HSL;
  }


  get transparent() {
    return this.alpha === 0 ? 'transparent' : '';
  }

  get keyword() {
    return hexadecimal2Keyword[this._sixDightHexadecimal];
  }
}